public/app/plugins/datasource/cloudwatch/query-runner/CloudWatchLogsQueryRunner.test.ts

```diff
@@ -0,0 +1,325 @@
+
+      expect(logGroups).toEqual(expectedLogGroups);
+    });
+  });
+
+  describe('getLogRowContext', () => {
+    it('replaces parameters correctly in the query', async () => {
+      const { runner, fetchMock } = setupMockedLogsQueryRunner();
+      const row: LogRowModel = {
+        entryFieldIndex: 0,
+        rowIndex: 0,
+        dataFrame: new MutableDataFrame({
+          refId: 'B',
+          fields: [
+            { name: 'ts', type: FieldType.time, values: [1] },
+            { name: LOG_IDENTIFIER_INTERNAL, type: FieldType.string, values: ['foo'], labels: {} },
+            { name: LOGSTREAM_IDENTIFIER_INTERNAL, type: FieldType.string, values: ['bar'], labels: {} },
+          ],
+        }),
+        entry: '4',
+        labels: {},
+        hasAnsi: false,
+        hasUnescapedContent: false,
+        raw: '4',
+        logLevel: LogLevel.info,
+        timeEpochMs: 4,
+        timeEpochNs: '4000000',
+        timeFromNow: '',
+        timeLocal: '',
+        timeUtc: '',
+        uid: '1',
+      };
+      await runner.getLogRowContext(row);
+      expect(fetchMock.mock.calls[0][0].data.queries[0].endTime).toBe(4);
+      expect(fetchMock.mock.calls[0][0].data.queries[0].region).toBe(undefined);
+
+      await runner.getLogRowContext(row, { direction: 'FORWARD' }, { ...validLogsQuery, region: 'eu-east' });
+      expect(fetchMock.mock.calls[1][0].data.queries[0].startTime).toBe(4);
+      expect(fetchMock.mock.calls[1][0].data.queries[0].region).toBe('eu-east');
+    });
+  });
+
+  describe('getLogGroupFields', () => {
+    it('passes region correctly', async () => {
+      const { runner, fetchMock } = setupMockedLogsQueryRunner();
+      fetchMock.mockReturnValueOnce(
+        of({
+          data: {
+            results: {
+              A: {
+                frames: [
+                  dataFrameToJSON(
+                    new MutableDataFrame({
+                      fields: [
+                        { name: 'key', values: [] },
+                        { name: 'val', values: [] },
+                      ],
+                    })
+                  ),
+                ],
+              },
+            },
+          },
+        })
+      );
+      await runner.getLogGroupFields({ region: 'us-west-1', logGroupName: 'test' });
+      expect(fetchMock.mock.calls[0][0].data.queries[0].region).toBe('us-west-1');
+    });
+  });
+
+  describe('logs query', () => {
+    beforeEach(() => {
+      jest.spyOn(rxjsUtils, 'increasingInterval').mockImplementation(() => interval(100));
+    });
+
+    it('should stop querying when timed out', async () => {
+      const { runner } = setupMockedLogsQueryRunner();
+      const fakeFrames = genMockFrames(20);
+      const initialRecordsMatched = fakeFrames[0].meta!.stats!.find((stat) => stat.displayName === 'Records scanned')!
+        .value!;
+      for (let i = 1; i < 4; i++) {
+        fakeFrames[i].meta!.stats = [
```

Comment:
seems sort of weird we need a ! for mocks right? Should we know what it contains?

Reply:
Yeah this looks weird to me too, but same here - would like to not change tests in this PR.